src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,594 @@
+    return struct {
+        allocator: Allocator,
+        env: *c.MDB_env,
+        dbis: []const c.MDB_dbi,
+        path: []const u8,
+
+        const Self = @This();
+
+        pub fn open(allocator: Allocator, logger: Logger, path: []const u8) anyerror!Self {
+            logger.info().log("Initializing LMDB");
+            const owned_path = try allocator.dupe(u8, path);
+
+            // create and open the database
+            const env = try ret(c.mdb_env_create, .{});
+            try result(c.mdb_env_set_maxdbs(env, column_families.len));
+            try result(c.mdb_env_open(env, @ptrCast(path), 0, 0o700));
+
+            // begin transaction to create column families aka "databases" in lmdb
+            const txn = try ret(c.mdb_txn_begin, .{ env, null, 0 });
+            errdefer c.mdb_txn_abort(txn);
+
+            // allocate cf handles
+            const dbis = try allocator.alloc(c.MDB_dbi, column_families.len);
+            errdefer allocator.free(dbis);
+
+            // save cf handles
+            inline for (column_families, 0..) |cf, i| {
+                // open cf/database, creating if necessary
+                dbis[i] = try ret(c.mdb_dbi_open, .{
+                    txn,
+                    @as([*]const u8, @ptrCast(cf.name)),
+                    c.MDB_CREATE,
+                });
+            }
+
+            // persist column families
+            try result(c.mdb_txn_commit(txn));
+
+            return .{
+                .allocator = allocator,
+                .env = env,
+                .dbis = dbis,
+                .path = owned_path,
+            };
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.allocator.free(self.dbis);
+            self.allocator.free(self.path);
+        }
+
+        fn dbi(self: *Self, comptime cf: ColumnFamily) c.MDB_dbi {
+            return self.dbis[cf.find(column_families)];
+        }
+
+        pub fn count(self: *Self, comptime cf: ColumnFamily) LmdbOrAllocatorError!u64 {
+            const txn = try ret(c.mdb_txn_begin, .{ self.env, null, c.MDB_RDONLY });
+            defer c.mdb_txn_abort(txn);
+
+            const stat = try ret(c.mdb_stat, .{ txn, self.dbi(cf) });
+
+            return stat.ms_entries;
+        }
+
+        pub fn put(
+            self: *Self,
+            comptime cf: ColumnFamily,
+            key: cf.Key,
+            value: cf.Value,
+        ) anyerror!void {
+            const key_bytes = try key_serializer.serializeToRef(self.allocator, key);
+            defer key_bytes.deinit();
+            const val_bytes = try value_serializer.serializeToRef(self.allocator, value);
+            defer val_bytes.deinit();
+
+            const txn = try ret(c.mdb_txn_begin, .{ self.env, null, 0 });
+            errdefer c.mdb_txn_abort(txn);
+
+            var key_val = toVal(key_bytes.data);
+            var val_val = toVal(val_bytes.data);
+            try result(c.mdb_put(txn, self.dbi(cf), &key_val, &val_val, 0));
```

Comment:
whats the expected behaviour regarding putting key that already exist? Will it override, error or append and does the `0` flag in `c.mdb_put` aligns with the expected behaviour?

Reply:
The default is to replace the existing data. 0 just means no flags, so use the default.

```c
	/** @brief Store items into a database.
	 *
	 * This function stores key/data pairs in the database. The default behavior
	 * is to enter the new key/data pair, replacing any previously existing key
	 * if duplicates are disallowed, or adding a duplicate data item if
	 * duplicates are allowed (#MDB_DUPSORT).
	 * @param[in] txn A transaction handle returned by #mdb_txn_begin()
	 * @param[in] dbi A database handle returned by #mdb_dbi_open()
	 * @param[in] key The key to store in the database
	 * @param[in,out] data The data to store
	 * @param[in] flags Special options for this operation. This parameter
	 * must be set to 0 or by bitwise OR'ing together one or more of the
	 * values described here.
	 * <ul>
	 *	<li>#MDB_NODUPDATA - enter the new key/data pair only if it does not
	 *		already appear in the database. This flag may only be specified
	 *		if the database was opened with #MDB_DUPSORT. The function will
	 *		return #MDB_KEYEXIST if the key/data pair already appears in the
	 *		database.
	 *	<li>#MDB_NOOVERWRITE - enter the new key/data pair only if the key
	 *		does not already appear in the database. The function will return
	 *		#MDB_KEYEXIST if the key already appears in the database, even if
	 *		the database supports duplicates (#MDB_DUPSORT). The \b data
	 *		parameter will be set to point to the existing item.
	 *	<li>#MDB_RESERVE - reserve space for data of the given size, but
	 *		don't copy the given data. Instead, return a pointer to the
	 *		reserved space, which the caller can fill in later - before
	 *		the next update operation or the transaction ends. This saves
	 *		an extra memcpy if the data is being generated later.
	 *		LMDB does nothing else with this memory, the caller is expected
	 *		to modify all of the space requested. This flag must not be
	 *		specified if the database was opened with #MDB_DUPSORT.
	 *	<li>#MDB_APPEND - append the given key/data pair to the end of the
	 *		database. This option allows fast bulk loading when keys are
	 *		already known to be in the correct order. Loading unsorted keys
	 *		with this flag will cause a #MDB_KEYEXIST error.
	 *	<li>#MDB_APPENDDUP - as above, but for sorted dup data.
	 * </ul>
	 * @return A non-zero error value on failure and 0 on success. Some possible
	 * errors are:
	 * <ul>
	 *	<li>#MDB_MAP_FULL - the database is full, see #mdb_env_set_mapsize().
	 *	<li>#MDB_TXN_FULL - the transaction has too many dirty pages.
	 *	<li>EACCES - an attempt was made to write in a read-only transaction.
	 *	<li>EINVAL - an invalid parameter was specified.
	 * </ul>
	 */
int  mdb_put(MDB_txn *txn, MDB_dbi dbi, MDB_val *key, MDB_val *data,
			    unsigned int flags);
```

https://github.com/LMDB/lmdb/blob/da9aeda08c3ff710a0d47d61a079f5a905b0a10a/libraries/liblmdb/lmdb.h#L1335-L1383